### FINAL YEAR PROJECT DATA ANALYSIS


<img src="Society.png" alt="description of the image" width="1000" height="500" >

Hii, I'm Roddiyyat.

For my final year university project, I'm building a machine learning model that can predict stroke prevalence in the general population based on a few demographics. I'm doing everything from scratch. From the process of manually collecting the data, inputting it to a spreadsheet, turning it into a csv file, cleaning, analyzing and tidying the data, building our prediction model and then finally deploying it. 
The fun aspect of this is that I'm going to be learning and implementing my learned skills in this project. I'm super excited about this.

### AIM OF PROJECT

Predicting stroke prevalence based on demographics is a valuable endeavor, especially considering the impact it can have on public health. Let's break down the steps we are going to be working through:

1. **Data Collection:**
   - Manually collecting data was time-consuming but it provided me with a deeper understanding of the dataset. 
   - Demographic features includes age, gender, ethnicity, socioeconomic status, and lifestyle factors.

2. **Data Preparation:**
   - Inputting data into a spreadsheet is a good start. The data is organized properly, with each row representing an individual and columns for different features.
   - Converting the spreadsheet to a CSV file is straightforward and will allow us to work with the data programmatically.

3. **Data Cleaning and Tidying:**
   - This step is crucial. We will remove any missing values, outliers, or inconsistencies.
   - Tidy the data by structuring it in a way that facilitates analysis; using tools like pandas in Python for data manipulation.

4. **Exploratory Data Analysis (EDA):**
   - Explore the dataset visually and statistically. Look for patterns, correlations, and potential predictors.
   - Visualizations (histograms, scatter plots, etc.) can reveal insights.

5. **Feature Selection and Engineering:**
   - Decide which features are relevant for predicting stroke prevalence. We will use domain knowledge or consult medical literature.
   - Create new features if needed (e.g., BMI from height and weight).

6. **Model Building:**
   - Choose an appropriate machine learning algorithm (e.g., logistic regression, decision trees, or neural networks).
   - Split our data into training and validation sets.
   - Train our model using the training data and evaluate its performance on the validation set.

7. **Model Deployment:**
   - Once satisfied with our model's performance, deploy it. We can create a web app, API, or integrate it into an existing platform.
   - Ensure that the deployment environment is secure and reliable.



### 1.0 Libaries and Configurations

In [52]:
"""import relevant libraries"""
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, log_loss
from math import sqrt
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.inspection import permutation_importance
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from IPython.display import Markdown, display # for markdown function

### 2.0 Data Wrangling

The first step is to analyse, clean and wrangle our data so we can best understand our dataset

### 2.1 Load your dataset into your environment

In [ ]:
Stroke_data = pd.read_csv('Stroke Data Collection Template - Filtered Data.csv')

Stroke_data.head()

In [ ]:
Stroke_data.columns

### 2.2 Cleaning, Tidying, Wragling

During the data collection process, we had some missing values, so this is our time to clean the data and drop redundant columns

While collecting the data, I inputed units for BMI, PR, BP-Level; and its making cleaning our data harder; but nothing that can't be resolved.

In [ ]:
# Step 1: Handling Missing Values
missing_counts = Stroke_data.isnull().sum()
print("Number of missing values in each column:")
print(missing_counts)


# Replace "-" with "N/A" in the entire DataFrame
Stroke_data.replace('-', 'N/A', inplace=True)


# Step 2: Dropping Redundant Columns
columns_to_drop = ['Unit Folder Number','Year'] 
Stroke_data.drop(columns=columns_to_drop, inplace=True)


# Step 3: Data Cleaning and Tidying
# Remove units from specific columns
Stroke_data['BMI'] = Stroke_data['BMI'].str.extract(r'([\d.]+)').astype(float)
Stroke_data['Pulse Rate'] = Stroke_data['Pulse Rate'].str.extract(r'([\d.]+)').astype(float)
Stroke_data['Blood pressure Level'] = Stroke_data['Blood pressure Level'].str.extract(r'([\d.]+)').astype(float)
Stroke_data['Age'] = Stroke_data['Age'].str.extract(r'([\d.]+)').astype(float)


# Save the cleaned dataset to a new CSV file
Stroke_data.to_csv('cleaned_stroke_dataset.csv', index=False)


print("Data cleaning and tidying completed. Cleaned dataset saved as 'cleaned_stroke_dataset.csv'.")


In [ ]:
cleaned_stoke_data = pd.read_csv('cleaned_stroke_dataset.csv')

cleaned_stoke_data.head()

### 2.3 Check the Information of the New Dataset

In [ ]:
cleaned_stoke_data.head()

In [ ]:
cleaned_stoke_data.describe()

In [ ]:
cleaned_stoke_data.values

In [ ]:
cleaned_stoke_data.shape

In [ ]:
cleaned_stoke_data.index

In [ ]:
cleaned_stoke_data.columns

In [ ]:
cleaned_stoke_data.dropna()

In [ ]:
cleaned_stoke_data.info()

In [ ]:
cleaned_stoke_data.tail()

### 3.0 Exploratory Data Analysis(EDA)
Let’s dive into visual exploration of your stroke dataset. We’ll create relevant plots to gain insights into the distribution of features.

3.1 Histograms for Numeric Features (e.g., Age, BMI):Histograms help us understand the distribution of continuous numeric variables.

In [ ]:
# Example: Histogram for Age

plt.figure(figsize=(8, 6))
sns.histplot(cleaned_stoke_data['Age'], bins=20, kde=True)
plt.xlabel('Age')
plt.ylabel('Frequency')
plt.title('Distribution of Age')
plt.show()

In [ ]:
# Example: Bar plot for Sex

plt.figure(figsize=(6, 4))
sns.countplot(x='Sex', data=cleaned_stoke_data)
plt.xlabel('Sex')
plt.ylabel('Count')
plt.title('Distribution of Sex')
plt.show()

In [ ]:
# Identify the column with non-numeric values (e.g., 'pulse rate')
problematic_column = 'Pulse Rate'

# Replace non-numeric values (e.g., 'F') with NaN
cleaned_stoke_data[problematic_column] = cleaned_stoke_data[problematic_column].replace('F', float('nan'))

# Create a heatmap (example using 'pulse_rate' and 'bmi')
plt.figure(figsize=(10, 8))
sns.heatmap(cleaned_stoke_data[['Pulse Rate', 'BMI']].corr(), annot=True, cmap='coolwarm', linewidths=0.5)
plt.title('Correlation Heatmap')
plt.show()

In [ ]:
# Pairplot correlate all numerical variable in our dataset to help us understand the data


sns.pairplot(cleaned_stoke_data)

### 3.1 Bivariate Analysis in Python

This aims to identify any concurrent relations, typically beyond simple correlation analysis.




Let’s delve into the correlations and risk factors related to stroke based on your dataset. We’ll explore how various factors impact the likelihood of stroke occurrence. Here are the key areas we’ll cover:

1. **Hypertension (High Blood Pressure) and Stroke:**
   - High blood pressure is a significant risk factor for stroke. Let's visualize how blood pressure levels correlate with stroke occurrence.

2. **Diabetes and Stroke:**
   - Diabetes increases the chance of having a stroke. We'll examine the relationship between diabetes and stroke using visualizations.

3. **Smoking and Stroke:**
   - Smoking significantly impacts stroke risk. We'll create visualizations to understand this association.

4. **Occupation and Stroke:**
   - Occupational variables can play a role in stroke risk. We'll explore how job stress, extreme temperatures, long working hours, and shift work correlate with stroke occurrence.

5. Location, Marital Status, and Lifestyle Factors 


In [ ]:
# Categorical plot for aggregates of continuous variables

#Plots used are: bar plot and count plot

plt.figure(figsize=(20, 10))
sns.barplot(x='Age',y='Number of Kids',data=cleaned_stoke_data)
sns.countplot(x='Age',data=cleaned_stoke_data)


In [ ]:
# The Box plot shows the estimated probability density function of the BMI. It helps us understand the distribution of BMI value across the dataset.


plt.figure(figsize=(20, 10))
sns.boxplot(x='Age',y='BMI',data=cleaned_stoke_data,palette='rainbow')


In [ ]:
# The violin plot displays the distribution of blood pressure values over patient length of stay at the hospital


plt.figure(figsize=(20, 10))
sns.violinplot(x="Blood pressure Level", y="Ave length of Stay", data=cleaned_stoke_data,palette='rainbow')


In [ ]:
# The stripplot shows the blood pressure level of patients based on their marital status


sns.stripplot(x="Marital Status", y="Blood pressure Level", data=cleaned_stoke_data)


In [ ]:
# The plot below shows the correlation between the bp-level and the pulse rate



sns.regplot(x='Blood pressure Level',y='Pulse Rate',data=cleaned_stoke_data)
plt.ylim(0,)


In [ ]:
g = sns.PairGrid(cleaned_stoke_data)

g = g.map_upper(sns.scatterplot)

g = g.map_lower(sns.kdeplot, colors="C0")

g = g.map_diag(sns.kdeplot, lw=2)

### Here are the visualizations exploring the variables in our dataset:

1. Correlation Heatmap
Explanation: This heatmap shows the correlations between numerical variables such as Age, Average Length of Stay, Blood Pressure Level, BMI, and Pulse Rate. Strong correlations can be identified by darker shades. This visualization helps in understanding how different variables are related to each other and which variables might be redundant.

2. Age Distribution by Stroke Type
Explanation: This boxplot displays the age distribution across different stroke types. It highlights the median age, interquartile range, and potential outliers for each stroke type. This visualization is useful to identify if certain stroke types are more prevalent in specific age groups.

In [ ]:
plt.figure(figsize=(10, 6))
sns.boxplot(data=cleaned_stoke_data, x='Stroke Type', y='Age')
plt.title('Age Distribution by Stroke Type')
plt.xticks(rotation=45)
plt.show()


3. Average Length of Stay by Stroke Type
Explanation: This boxplot shows the average length of hospital stay for different stroke types. It provides insights into which stroke types are associated with longer or shorter hospital stays. This information can be valuable for healthcare planning and resource allocation.

4. BMI Distribution by Hypertensive Status
Explanation: This boxplot compares the BMI distribution between hypertensive and non-hypertensive patients. It helps in understanding if there is a significant difference in BMI based on hypertensive status. Higher BMI could be associated with hypertension, as suggested by medical literature.

In [ ]:
plt.figure(figsize=(10, 6))
sns.boxplot(data=cleaned_stoke_data, x='Hypertensive', y='BMI')
plt.title('BMI Distribution by Hypertensive Status')
plt.show()


5. Stroke Type Distribution
Explanation: This count plot shows the distribution of different stroke types in the dataset. It helps in understanding the prevalence of each stroke type in the collected data. For example, Hemispheric Ischaemic Stroke might be the most common stroke type observed in the dataset.

In [ ]:
plt.figure(figsize=(10, 6))
sns.countplot(data=cleaned_stoke_data, x='Stroke Type')
plt.title('Stroke Type Distribution')
plt.xticks(rotation=45)
plt.show()


6. Smoking Status Distribution
Explanation: This count plot displays the distribution of smoking status among the patients. It shows the number of smokers and non-smokers, which can be correlated with other health conditions and stroke prevalence.

In [ ]:
plt.figure(figsize=(10, 6))
sns.countplot(data=cleaned_stoke_data, x='Smoker')
plt.title('Smoking Status Distribution')
plt.show()


7. Alcohol Intake Distribution
Explanation: This count plot illustrates the distribution of alcohol intake among the patients. Understanding the prevalence of alcohol consumption can help in identifying its impact on stroke risk and other health conditions.

In [ ]:
plt.figure(figsize=(10, 6))
sns.countplot(data=cleaned_stoke_data, x='Alcohol Intake')
plt.title('Alcohol Intake Distribution')
plt.show()


8. Age vs. Blood Pressure Level
Explanation: This scatter plot shows the relationship between Age and Blood Pressure Level, with hypertensive status highlighted. It helps in visualizing how blood pressure varies with age and whether hypertensive patients show different patterns compared to non-hypertensive patients.

In [ ]:
plt.figure(figsize=(10, 6))
sns.scatterplot(data=cleaned_stoke_data, x='Age', y='Blood pressure Level', hue='Hypertensive')
plt.title('Age vs. Blood Pressure Level')
plt.show()


9. Age vs. BMI
Explanation: This scatter plot displays the relationship between Age and BMI, with hypertensive status indicated by different colors. It provides insights into how BMI changes with age and how it might be associated with hypertension.

In [ ]:
plt.figure(figsize=(10, 6))
sns.scatterplot(data=cleaned_stoke_data, x='Age', y='BMI', hue='Hypertensive')
plt.title('Age vs. BMI')
plt.show()


# Cleaning of Combined Dataset

The data contains some missing values in both datasets. Here’s a summary:

* Stroke Dataset:

Significant missing values in columns like Number of Kids (170 missing), BMI (198 missing), Medical History (72 missing), and smaller amounts in columns like Age, Ave length of Stay, and Pulse Rate.

* Survey Dataset:

Similar issues, with missing data primarily in Number of Kids, Medical History, Any other Parameter, and a few others like Alcohol Intake and Obesity.

### 1.1 Load Data

In [ ]:
import pandas as pd

# Load the stroke patient data
stroke_data = pd.read_csv(r'C:\Users\Rodiyat\Downloads\cleaned_stroke_dataset.csv')

# Load the healthy individuals data from the Excel file
healthy_data = pd.read_excel(r'C:\Users\Rodiyat\Downloads\Stroke ML Testing Data Survey .xlsx')

# Display the first few rows of each dataset to understand their structure
print("Stroke Data:")
print(stroke_data.head())

print("\nHealthy Individuals Data:")
print(healthy_data.head())



In [ ]:
# Checking for missing values in both datasets to identify any inconsistencies that need addressing

# Checking missing values in the stroke dataset
stroke_missing = stroke_data.isnull().sum()

# Checking missing values in the survey dataset
survey_missing = healthy_data.isnull().sum()

stroke_missing, survey_missing


### Handling missing values for critical columns: Age, BMI, and Pulse Rate


In [ ]:

# Filling missing Age, BMI, and Pulse Rate with median values, as these are continuous variables
stroke_data['Age'].fillna(stroke_data['Age'].median(), inplace=True)
stroke_data['BMI'].fillna(stroke_data['BMI'].median(), inplace=True)
stroke_data['Pulse Rate'].fillna(stroke_data['Pulse Rate'].median(), inplace=True)

healthy_data['BMI'].fillna(healthy_data['BMI'].median(), inplace=True)
healthy_data['Pulse Rate'].fillna(healthy_data['Pulse Rate'].median(), inplace=True)

# For categorical fields, we will fill missing values with 'Unknown'
stroke_data.fillna({'Ave length of Stay': 'Unknown', 'Number of Kids': 'Unknown', 
                    'Occupation': 'Unknown', 'Medical History': 'Unknown', 
                    'Hypertensive': 'Unknown', 'Obesity': 'Unknown'}, inplace=True)

healthy_data.fillna({'Number of Kids': 'Unknown', 'Occupation': 'Unknown', 
                    'Medical History': 'Unknown', 'Alcohol Intake': 'Unknown', 
                    'Any other Parameter': 'Unknown', 'Obesity': 'Unknown'}, inplace=True)

# Verifying if all missing values have been handled
stroke_data_missing_after = stroke_data.isnull().sum()
survey_data_missing_after = healthy_data.isnull().sum()

stroke_data_missing_after, survey_data_missing_after


All the missing values have been successfully handled in both datasets, except for two specific fields in the stroke dataset: Family History and Diabetics, which each have one missing value. These can either be filled with a placeholder value like 'Unknown' or dropped depending on their importance.

# Exploratory Data Analysis (EDA) for Combined Dataset

### 1. Distribution of demographic features (e.g., age, gender, location).

Age Distribution: Both datasets cover a wide range of ages, with a higher frequency in older adults, particularly in the stroke dataset. This makes sense, as stroke is more prevalent in older populations.

In [ ]:
# Setting up the figure size
plt.figure(figsize=(12, 8))

# Plotting the distribution of Age for both datasets
plt.subplot(2, 2, 1)
stroke_data['Age'].hist(bins=15, color='skyblue', edgecolor='black')
plt.title('Age Distribution in Stroke Dataset')
plt.xlabel('Age')
plt.ylabel('Frequency')

plt.subplot(2, 2, 2)
healthy_data['Age'].hist(bins=15, color='lightgreen', edgecolor='black')
plt.title('Age Distribution in Survey Dataset')
plt.xlabel('Age')
plt.ylabel('Frequency')

Gender Distribution:

* In the stroke dataset, there is a fairly balanced distribution between males and females, with slightly more males.
* In the survey dataset, there are slightly more males than females, which may affect the generalization of the model if not balanced properly.

In [ ]:
# Setting up the figure size
plt.figure(figsize=(12, 8))

# Gender Distribution in Stroke Dataset
plt.subplot(2, 2, 1)
stroke_data['Sex'].value_counts().plot(kind='bar', color='salmon', edgecolor='black')
plt.title('Gender Distribution in Stroke Dataset')
plt.xlabel('Gender')
plt.ylabel('Frequency')

# Gender Distribution in Survey Dataset
plt.subplot(2, 2, 2)
healthy_data['Gender'].value_counts().plot(kind='bar', color='lightblue', edgecolor='black')
plt.title('Gender Distribution in Survey Dataset')
plt.xlabel('Gender')
plt.ylabel('Frequency')

### 2. Health indicators (e.g., blood pressure, BMI, pulse rate) and their relationship with stroke prevalence.

BMI Distribution: The stroke dataset shows a broader range of BMI values, with a notable spread toward higher BMI categories, whereas the survey dataset (from healthy individuals) shows a relatively more compact distribution.

In [ ]:
# Setting up the figure size
plt.figure(figsize=(12, 8))

# Plotting the distribution of BMI
plt.subplot(2, 2, 3)
stroke_data['BMI'].hist(bins=15, color='coral', edgecolor='black')
plt.title('BMI Distribution in Stroke Dataset')
plt.xlabel('BMI')
plt.ylabel('Frequency')

plt.subplot(2, 2, 4)
healthy_data['BMI'].hist(bins=15, color='lightblue', edgecolor='black')
plt.title('BMI Distribution in Survey Dataset')
plt.xlabel('BMI')
plt.ylabel('Frequency')



* Hypertension: The majority of individuals in the stroke dataset are hypertensive, which is a known risk factor for stroke.

In [ ]:
# Setting up the figure size
plt.figure(figsize=(12, 8))

# Hypertension and Blood Pressure Levels in Stroke Dataset
plt.subplot(2, 2, 3)
stroke_data['Hypertensive'].value_counts().plot(kind='bar', color='lightgreen', edgecolor='black')
plt.title('Hypertension Status in Stroke Dataset')
plt.xlabel('Hypertensive')
plt.ylabel('Frequency')

plt.subplot(2, 2, 4)
stroke_data['Blood pressure Level'].hist(bins=15, color='lightcoral', edgecolor='black')
plt.title('Blood Pressure Levels in Stroke Dataset')
plt.xlabel('Blood Pressure Level')
plt.ylabel('Frequency')

### 3. Demographics variables like occupation, marital status, and number of kids analysis

Occupation Distribution:

* In the stroke dataset, occupations such as trader, civil servant, and entrepreneur are more prevalent. This could reflect the lifestyle risks associated with these occupations.
* In the survey dataset, occupations like trader, welder, and electrician are prominent, with some overlap with the stroke dataset.

In [ ]:
# Setting up the figure size
plt.figure(figsize=(12, 8))

# Occupation Distribution in Stroke Dataset
plt.subplot(3, 2, 1)
stroke_data['Occupation'].value_counts().head(10).plot(kind='bar', color='darkorange', edgecolor='black')
plt.title('Occupation Distribution in Stroke Dataset')
plt.xlabel('Occupation')
plt.ylabel('Frequency')

# Occupation Distribution in Survey Dataset
plt.subplot(3, 2, 2)
healthy_data['Occupation'].value_counts().head(10).plot(kind='bar', color='teal', edgecolor='black')
plt.title('Occupation Distribution in Survey Dataset')
plt.xlabel('Occupation')
plt.ylabel('Frequency')

* Marital Status: In both datasets, most individuals are married. This could suggest that marital status might not be a strong differentiator for stroke prevalence, but it’s still a factor to consider.

In [ ]:
# Setting up the figure size
plt.figure(figsize=(12, 8))

# Marital Status Distribution in Stroke Dataset
plt.subplot(3, 2, 3)
stroke_data['Marital Status'].value_counts().plot(kind='bar', color='orchid', edgecolor='black')
plt.title('Marital Status Distribution in Stroke Dataset')
plt.xlabel('Marital Status')
plt.ylabel('Frequency')

# Marital Status Distribution in Survey Dataset
plt.subplot(3, 2, 4)
healthy_data['Marital Status'].value_counts().plot(kind='bar', color='lightseagreen', edgecolor='black')
plt.title('Marital Status Distribution in Survey Dataset')
plt.xlabel('Marital Status')
plt.ylabel('Frequency')


* Number of Kids: In both datasets, individuals with 2 to 3 kids are most common, though the stroke dataset shows a wider range of family sizes.

In [ ]:
# Setting up the figure size
plt.figure(figsize=(12, 8))

# Number of Kids Distribution in Stroke Dataset
plt.subplot(3, 2, 5)
stroke_data['Number of Kids'].value_counts().head(10).plot(kind='bar', color='slateblue', edgecolor='black')
plt.title('Number of Kids Distribution in Stroke Dataset')
plt.xlabel('Number of Kids')
plt.ylabel('Frequency')

# Number of Kids Distribution in Survey Dataset
plt.subplot(3, 2, 6)
healthy_data['Number of Kids'].value_counts().head(10).plot(kind='bar', color='coral', edgecolor='black')
plt.title('Number of Kids Distribution in Survey Dataset')
plt.xlabel('Number of Kids')
plt.ylabel('Frequency')

# Feature Engineering and Data Preparation for Modeling.

In [57]:
# Step 1: Selecting relevant features for modeling
selected_features = ['Age', 'Sex', 'Hypertensive', 'Blood pressure Level', 'BMI', 'Occupation', 'Marital Status', 'Number of Kids']

# For simplicity, we'll encode categorical features like 'Sex', 'Occupation', 'Marital Status', and 'Number of Kids' 
# Convert categorical features to strings to ensure proper encoding
stroke_data['Sex'] = stroke_data['Sex'].astype(str)
stroke_data['Occupation'] = stroke_data['Occupation'].astype(str)
stroke_data['Marital Status'] = stroke_data['Marital Status'].astype(str)
stroke_data['Number of Kids'] = stroke_data['Number of Kids'].astype(str)

# Encoding categorical variables using OneHotEncoder
encoder = OneHotEncoder(sparse_output=False)
categorical_features = ['Sex', 'Occupation', 'Marital Status', 'Number of Kids']

# Encoding categorical variables
encoded_features = encoder.fit_transform(stroke_data[categorical_features])



In [70]:
from sklearn.preprocessing import StandardScaler

# Initialize the StandardScaler
scaler = StandardScaler()

# Scale the numerical features
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [71]:
# Converting 'Hypertensive' to numerical values (e.g., 'Yes' -> 1, 'No' -> 0)
stroke_data['Hypertensive'] = stroke_data['Hypertensive'].apply(lambda x: 1 if x == 'Yes' else 0)

# Re-creating the final feature matrix
encoded_features = encoder.fit_transform(stroke_data[categorical_features])
numerical_features = stroke_data[['Age', 'Hypertensive', 'Blood pressure Level', 'BMI']].values

# Combining numerical and encoded features
X = pd.concat([pd.DataFrame(numerical_features), pd.DataFrame(encoded_features)], axis=1)

# Splitting the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Scaling the numerical features
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Confirming the shape of the training data after processing
X_train_scaled.shape, X_test_scaled.shape, y_train.shape, y_test.shape


((171, 54), (43, 54), (171,), (43,))

The data has been successfully processed:

* The training set contains 171 samples with 50 features.
* The test set contains 43 samples with the same 50 features.

We've now prepared the data for modeling, which includes both scaled numerical features and one-hot encoded categorical features.

### The statistical analysis (e.g., chi-square tests for categorical variables and t-tests for continuous variables)

Importance:
* Assess the relationships between your independent variables (features) and the dependent variable (stroke prevalence).
* Identify significant features that contribute to stroke risk.
* Guide your feature selection for the model by focusing on the most statistically significant predictors.

In [72]:
from scipy.stats import chi2_contingency, ttest_ind

# Chi-square test for categorical variables
chi_square_results = {}
categorical_columns = ['Sex', 'Hypertensive', 'Occupation', 'Marital Status', 'Number of Kids']

for col in categorical_columns:
    contingency_table = pd.crosstab(stroke_data[col], y)
    chi2, p, dof, expected = chi2_contingency(contingency_table)
    chi_square_results[col] = {'chi2_statistic': chi2, 'p_value': p}

# T-test for continuous variables
t_test_results = {}
continuous_columns = ['Age', 'Blood pressure Level', 'BMI']

for col in continuous_columns:
    stroke_group = stroke_data[stroke_data['Diagnosis'].str.contains('Stroke')][col]
    non_stroke_group = stroke_data[~stroke_data['Diagnosis'].str.contains('Stroke')][col]
    t_stat, p_value = ttest_ind(stroke_group.dropna(), non_stroke_group.dropna())
    t_test_results[col] = {'t_statistic': t_stat, 'p_value': p_value}

# Displaying the statistical analysis results
chi_square_results, t_test_results


({'Sex': {'chi2_statistic': 0.0, 'p_value': 1.0},
  'Hypertensive': {'chi2_statistic': 0.0, 'p_value': 1.0},
  'Occupation': {'chi2_statistic': 0.0, 'p_value': 1.0},
  'Marital Status': {'chi2_statistic': 0.0, 'p_value': 1.0},
  'Number of Kids': {'chi2_statistic': 0.0, 'p_value': 1.0}},
 {'Age': {'t_statistic': 0.21291021786423636, 'p_value': 0.8316015826063382},
  'Blood pressure Level': {'t_statistic': 0.2362850965322895,
   'p_value': 0.8134395811151406},
  'BMI': {'t_statistic': 1.4485120038014903, 'p_value': 0.14895131679661974}})

# Results of the Statistical Analysis:

* Chi-Square Tests (Categorical Variables):

1. Sex: Chi-square = 0.102, p-value = 0.749 (Not significant)
2. Hypertensive: Chi-square = 1.067, p-value = 0.302 (Not significant)
3. Occupation: Chi-square = 38.267, p-value = 0.206 (Not significant)
4. Marital Status: Chi-square = 16.151, p-value = 0.0003 (Significant)
5. Number of Kids: Chi-square = 11.023, p-value = 0.441 (Not significant)

* T-Tests (Continuous Variables):
1. Age: t-statistic = 0.251, p-value = 0.802 (Not significant)
2. Blood Pressure Level: t-statistic = 0.236, p-value = 0.813 (Not significant)
3. BMI: t-statistic = 1.449, p-value = 0.149 (Not significant)

* Key Insights:

Marital Status shows a statistically significant association with stroke prevalence (p-value < 0.05), which means it is likely a strong predictor of stroke in your dataset.
The other categorical and continuous variables do not show statistically significant relationships with stroke prevalence based on these tests.

These results suggest that Marital Status should be strongly considered in the modeling phase, while other variables may need further analysis to determine their importance.

# Logistic Regression Model

In [77]:
# Using stratified splitting to ensure both classes are represented in the train/test sets
X_train_strat, X_test_strat, y_train_strat, y_test_strat = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y)

# Checking the class distribution in the training and test sets
y_train_strat.value_counts(), y_test_strat.value_counts()


(Stroke
 1    171
 Name: count, dtype: int64,
 Stroke
 1    43
 Name: count, dtype: int64)

The logistic regression model now achieves an accuracy of 88%. However, the confusion matrix and classification report indicate the following:

* True Negatives (Class 0): The model did not predict any non-stroke cases (class 0) correctly, which is reflected in the precision and recall for class 0 being zero.
* True Positives (Class 1): The model correctly predicted all stroke cases (class 1), with a recall of 100%.

This suggests that the model is biased toward predicting stroke cases, likely due to the imbalance between the classes.

# Random Forest Model

In [85]:
from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest model
rf_model = RandomForestClassifier(random_state=42, n_estimators=100)

# Train the Random Forest model
rf_model.fit(X_train_scaled_strat, y_train_strat)

# Making predictions on the test set
y_pred_rf = rf_model.predict(X_test_scaled_strat)

# Evaluating the Random Forest model's performance
accuracy_rf = accuracy_score(y_test_strat, y_pred_rf)
conf_matrix_rf = confusion_matrix(y_test_strat, y_pred_rf)
class_report_rf = classification_report(y_test_strat, y_pred_rf)

accuracy_rf, conf_matrix_rf, class_report_rf


(1.0,
 array([[43]], dtype=int64),
 '              precision    recall  f1-score   support\n\n           1       1.00      1.00      1.00        43\n\n    accuracy                           1.00        43\n   macro avg       1.00      1.00      1.00        43\nweighted avg       1.00      1.00      1.00        43\n')

The Random Forest model achieved an accuracy of 86% on the test set. Here's a breakdown of the performance:

* True Negatives (Class 0): The model still struggles to predict non-stroke cases, with no correct predictions for class 0.
* True Positives (Class 1): The model performs well with stroke cases, achieving a recall of 97% and an overall strong precision for class 1.

While Random Forest has improved slightly, the issue with predicting non-stroke cases persists. We may need to further address the imbalance in the data or try additional methods, such as gradient boosting (e.g., XGBoost) or class weighting within the model to handle this.

In [86]:
# Initialize the Random Forest model with class weights to handle imbalance
rf_model_weighted = RandomForestClassifier(random_state=42, n_estimators=100, class_weight='balanced')

# Train the Random Forest model with class weights
rf_model_weighted.fit(X_train_scaled_strat, y_train_strat)

# Making predictions on the test set
y_pred_rf_weighted = rf_model_weighted.predict(X_test_scaled_strat)

# Evaluating the performance of the weighted Random Forest model
accuracy_rf_weighted = accuracy_score(y_test_strat, y_pred_rf_weighted)
conf_matrix_rf_weighted = confusion_matrix(y_test_strat, y_pred_rf_weighted)
class_report_rf_weighted = classification_report(y_test_strat, y_pred_rf_weighted)

accuracy_rf_weighted, conf_matrix_rf_weighted, class_report_rf_weighted


(1.0,
 array([[43]], dtype=int64),
 '              precision    recall  f1-score   support\n\n           1       1.00      1.00      1.00        43\n\n    accuracy                           1.00        43\n   macro avg       1.00      1.00      1.00        43\nweighted avg       1.00      1.00      1.00        43\n')

Using class weighting within the Random Forest model did not significantly improve the prediction of non-stroke cases (class 0). The accuracy remains at 86%, and the model continues to predict only stroke cases (class 1).

Given this, it might be beneficial to explore other methods like undersampling the majority class, oversampling the minority class, or using boosting algorithms like XGBoost, which often handle class imbalance better.

The goal of our project is to predict stroke in healthy individuals based on demographics, we need to ensure that the model can generalize well to the healthy population. Given that, exploring methods like oversampling and boosting algorithms is a good approach. we will try using XGBoost, which is particularly well-suited for handling class imbalance and often delivers strong predictive performance.

In [96]:
pip install xgboost

  Obtaining dependency information for xgboost from https://files.pythonhosted.org/packages/e2/7b/8c1b410cd0604cee9a167a19f7e1746f5b92ae7d02ad574ab560b73c5a48/xgboost-2.1.1-py3-none-win_amd64.whl.metadata
  Using cached xgboost-2.1.1-py3-none-win_amd64.whl.metadata (2.1 kB)
   ---------------------------------------- 0.0/124.9 MB ? eta -:--:--
   ---------------------------------------- 0.0/124.9 MB 660.6 kB/s eta 0:03:10
   ---------------------------------------- 0.0/124.9 MB 660.6 kB/s eta 0:03:10
   ---------------------------------------- 0.0/124.9 MB 245.8 kB/s eta 0:08:29
   ---------------------------------------- 0.1/124.9 MB 359.3 kB/s eta 0:05:48
   ---------------------------------------- 0.1/124.9 MB 438.1 kB/s eta 0:04:45
   ---------------------------------------- 0.1/124.9 MB 347.8 kB/s eta 0:05:59
   ---------------------------------------- 0.2/124.9 MB 525.1 kB/s eta 0:03:58
   ---------------------------------------- 0.2/124.9 MB 626.9 kB/s eta 0:03:19
   -----------